### PR TITLE
Optimize checksum computation

### LINKED ISSUE #16372: Faster checksum implementation
### What is the problem this feature will solve?

Astro.py currently takes several seconds to add a datasum to large fits files. In my tests using a 1.1GB fits file, where most of the data is in the primary HDU, it took ~2.5s to perform the following:

```python
hdul = fits.open('inversion_output/scan1807/inv_res_pre.fits')
hdul[0].add_datasum()
hdul[0].header
```

### Describe the desired outcome

While implementing this in Haskell, I used the c code in the FITS 4.0 spec, and adding a datasum to the same HDU was about 10x faster. 

Code is included here. There are small differences from the code in the FITS 4.0 spec, namely returning the resulting sum instead of passing in a pointer. 

```c

unsigned int checksum (
  unsigned char *buf,
  int length)
{

  unsigned int sum32 = 0;
  unsigned int hi, lo, hicarry, locarry, i;

  // Accumulate the sum of the high-order 16 bits and the */
  // low-order 16 bits of each 32-bit word, separately. */
  // The first byte in each pair is the most significant. */
  // This algorithm works on both big and little endian machines.
  hi = (sum32 >> 16);
  lo = sum32 & 0xFFFF;
  for (i=0; i < length; i+=4) {
    hi += ((buf[i] << 8) + buf[i+1]);
    lo += ((buf[i+2] << 8) + buf[i+3]);
  }

  // fold carry bits from each 16 bit sum into the other sum
  hicarry = hi >> 16;
  locarry = lo >> 16;

  while (hicarry || locarry) {
    hi = (hi & 0xFFFF) + locarry;
    lo = (lo & 0xFFFF) + hicarry;
    hicarry = hi >> 16;
    locarry = lo >> 16;
  }

  // Concatenate the high and low parts to form the full 32-bit checksum
  return (hi << 16) + lo;
}
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by saimn**
As suggested in #16372.

Closes #16372

Before:
```
In [1]: import numpy as np
   ...: from astropy.io import fits
   ...: rng = np.random.default_rng(0)
   ...: data = rng.integers(0, 2**16-1, (2048, 2048))
   ...: hdu = fits.ImageHDU(data=data)
   ...: 
   ...: %timeit hdu.add_datasum()
73.5 ms ± 579 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
After:
```
   ...: %timeit hdu.add_datasum()
7.98 ms ± 44.4 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Note that part of the speedup is due to removing a copy that seem unnecessary (`.flatten` → `.ravel`).

### Description

This pull request is to address ...

Fixes #<Issue Number>

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/io/fits/hdu/base.py [hunk 1]
```diff
@@ -1464,55 +1465,9 @@ def _compute_checksum(self, data, sum32=0):
...
-        This code should only be called with blocks of 2880 bytes
-        Longer blocks result in non-standard checksums with carry overflow
-        Historically,  this code *was* called with larger blocks and for that
-        reason still needs to be for backward compatibility.
-        """
-        u8 = np.uint32(8)
-        u16 = np.uint32(16)
-        uFFFF = np.uint32(0xFFFF)
-
-        if data.nbytes % 2:
```
Comment (neutrinoceros) [inline astropy/io/fits/hdu/base.py:1488]: this block didn't make it to the Cython version, is this intended ?

File: astropy/io/fits/hdu/base.py [hunk 2]
```diff
@@ -1464,55 +1465,9 @@ def _compute_checksum(self, data, sum32=0):
...
-            last = np.uint32(0)
-
-        data = data.view(">u2")
-
-        hi = sum32 >> u16
-        lo = sum32 & uFFFF
-        hi += np.add.reduce(data[0::2], dtype=np.uint64)
-        lo += np.add.reduce(data[1::2], dtype=np.uint64)
-
-        if (data.nbytes // 2) % 2:
```
Comment (neutrinoceros) [inline astropy/io/fits/hdu/base.py:1501]: here too

**Review by neutrinoceros — COMMENTED** [review]
failures look very real, though the cython code is (almost) identical to the Python version, so I'm wondering if differences are intended and whether they are the cause of a regression.

**Comment by mhvk** [comment]
Hmm, this whole routine felt like an example of coding when one does not have access to `long int` - and the whole lot can be replaced by a much simpler summation -- will push PR shortly...

**Comment by pllim** [comment]
Hmm regression under certain condition?

```
    def test_failing_compressed_datasum(self):
        """
        Regression test for https://github.com/astropy/astropy/issues/4587
        """
        n = np.ones((10, 10), dtype="float32")
        comp_hdu = fits.CompImageHDU(n)
        comp_hdu.writeto(self.temp("tmp.fits"), checksum=True)
    
>       with fits.open(self.temp("tmp.fits"), checksum=True) as hdul:
...
astropy.utils.exceptions.AstropyUserWarning: Checksum verification failed for HDU ('COMPRESSED_IMAGE', 1).
```

The other one seems to come from this in `test_connect.py` with `<_io.FileIO [closed]>`:

```python
    def test_memmap(self, tmp_path):
        filename = tmp_path / "test_simple.fts"
        t1 = Table(self.data)
        t1.write(filename, overwrite=True)
        t2 = Table.read(filename, memmap=False)
        t3 = Table.read(filename, memmap=True)
        assert equal_data(t2, t3)
        # To avoid issues with open files, we need to remove references to
        # data that uses memory mapping and force the garbage collection
        del t1, t2, t3
        gc.collect()
```

Anyway, I restarted the failing "double run" job, just in case.

**Comment by pllim** [comment]
Also the double run job uses the test runner we been thinking about getting rid of...

`double: python -c 'import sys; from astropy import test; test(); sys.exit(test())'`

**Comment by mhvk** [comment]
See #17209 for my alternative.

**Pull request closed without merging**